src/main/java/org/asamk/signal/Main.java

```diff
@@ -1293,6 +1296,59 @@ public void handleMessage(SignalServiceEnvelope envelope, SignalServiceContent c
         }
     }
 
+    private static class JsonDbusReceiveMessageHandler extends JsonReceiveMessageHandler {
+        final DBusConnection conn;
+
+        public JsonDbusReceiveMessageHandler(Manager m, DBusConnection conn) {
+            super(m);
+            this.conn = conn;
+        }
+
+        @Override
+        public void handleMessage(SignalServiceEnvelope envelope, SignalServiceContent content, Throwable exception) {
+            super.handleMessage(envelope, content, exception);
+
+            if (envelope.isReceipt()) {
```

Comment:
Move this complete `if` to a separate function, and also use that function from `DbusReceiveMessageHandler`, so the code isn't just copied.

Reply:
I think a lot of the trouble is due to the fact that Java is single inheritance; in a multiple inheritance language like Python, I could just make a JsonDbusHandler that inherits from both the JsonHandler and the DbusHandler and just have two super().handleMessage() calls, which would've made life a lot easier.

I think it might actually be better to figure out if there's a better way to architect this part entirely, but I'm not sure if this method is the best way to do it. I think it will definitely be good in the long run to generalise it more than this.

Reply:
One thing I was considering was possibly something like this: you have an array of callbacks, and whenever you receive a message, each callback gets called with the message and does stuff to it. Then you can write a general json print callback, dbus-message-sending callback, etc... (The specifics will obviously vary, and there are drawbaks, but this is just one way to do it off the top of my head).

Reply:
yeah, that'll need more thinking.